### DISCUSSION TITLE
SDRAM as heap not working in STM32H573

### DISCUSSION THREAD

**Discussion by kiranakd** [micropython/micropython#15515] [STM32 / Pyboard]
Hi All,

We have made custom board using STM32H573ZI. We have connected Alliance memory make SDRAM chip (AS4C1M16S-7TCN which has 16Mbit SDRAM) via FMC. We have tested it's connectivity in STM32CubeIDE which is working well. But it is not getting used in Micropython.
We have made all the necessary configurations (FMC pin assignments, SDRAM settings etc.) in mpconfigboard.h, added SDRAM section in linker script and built the firmware. When we don't assign the SDRAM address to heap, MCU boots well and firmware runs properly. But SDRAM will not get utilized in this case.

![image](https://github.com/user-attachments/assets/4fb8080b-9536-4d05-97a9-a3ec89d4c39b)

However, if we enable the heap assignment line, then firmware doesn't run / MCU doesn't boot.

![image](https://github.com/user-attachments/assets/107eddae-020b-47d5-affa-73a25b7cfd61)

below are the configuration settings we have made for SDRAM portion.

![image](https://github.com/user-attachments/assets/04f6f765-6ab1-4f11-962a-2076ccfefc0a)
![image](https://github.com/user-attachments/assets/5c4581c3-82d4-41c3-b688-56923ccd2d6a)

Looks like, something is missing in Micropython firmware modules related to STM32H5 series microcontrollers to use SDRAM as heap.
Please help us to resolve the above issue.

**Comment by kiranakd** [comment]
Hi,

Can anyone please provide inputs on this issue.
Thanks in advance.

**Comment by andrewleech** [comment]
Maybe take a look at https://github.com/micropython/micropython/blob/master/ports/stm32/mpu.h it possibly needs enabling and configuring for your chip?

**Comment by kiranakd** [comment]
Hi, Thanks for your input.

There is an another discussion similar to this which happened with H7 chip. 
https://github.com/micropython/micropython/issues/9087

They could resolve it by enabling mpu config for H7 chips in sdram.c file which is included by default in recent modules.
When we look at the code in mpu.h, configurations are different for H7 and H5. There is no code related to MPU_CONFIG_SDRAM and defined only MPU_REGION_SIG and MPU_REGION_ETH sections.

So not getting the required modifications for H5 chip. And couldn't get any info either.
Can you please provide the changes if you know.

Thanks.

**Reply by andrewleech** [reply]
I don't have any H5 parts so don't know exactly what would be needed sorry

**Reply by kiranakd** [reply]
It's OK sir.

Can anyone help me in this please.

**Comment by kiranakd** [comment]
Hi,

Can anyone please provide your valuable suggestions on this.

**Comment by kiranakd** [comment]
Hi,

Can anyone help on this?

**Comment by andrewleech** [comment]
Maybe look at the datasheets for the H5 and the h7 and compare the mpu sections. If they look pretty similar you might be able to mostly copy the sdram code section from h7 to H5 and see if that works.

**Reply by kiranakd** [reply]
Hi,

I had already tried copying the MPU_CONFIG_SDRAM portion for H5 in mpu.h and calling it in sdram.c file. But got errors during the build.
Actually, i have tried all things which i can do from my end before reaching this forum.
Looking for the exact code which can resolve the issue. Some how this portion is not done properly in Micropython firmware modules  for H5 series :(

**Reply by andrewleech** [reply]
There might not be anyone else with H5 hardware that has sdram. Usually around here support for this kind of thing is added by the first person who needs to use it.

**Reply by kiranakd** [reply]
Hi,

It is unfortunate to know that, people would have not integrated SDRAM with H5 in MPY so far.
Didn't get support from STM team also on this.

We will wait for some more time till we get support from either STM team or from any experts in this forum.

**Comment by kiranakd** [comment]
Hi,

Can anyone look into this issue please.

**Comment by kiranakd** [comment]
Hi,

Any hope, we can get solution for this?
